src/addresses.py

```
@@ -0,0 +1,103 @@
+    """
+    (postal_code_city, postal_code_state
+        ) = get_city_and_state_from_postal_code(country, postal_code)
+    return Address(
+        address = street_address,
+        city = city,
+        state = state,
+        postal_code = postal_code,
+        postal_code_city = postal_code_city,
+        postal_code_state = postal_code_state,
+        country=country.lower()
+    )
+
+def get_api_headers(api_name: str) -> Dict:
+    if api_name == 'ziptastic':
+        return { "x-key" : config.ZIPTASTIC_API_KEY }
+    if api_name == 'postcodes':
+        return {}
+    return {}
+
+def get_city_and_state_from_postal_code(
+    country_name: str, # the two letter code for the country
+    postal_code: str
+) -> Tuple[Optional[str], Optional[str]]:
+    """ Given a country code and a zip, fetch the city and state."""
+    standardized_country_code = country_name.upper()
+    if standardized_country_code == 'GB' or standardized_country_code == 'UK':
+        resp = requests.get(
+            BASE_POSTCODES_API_URL+postal_code,
+            headers = get_api_headers('postcodes')
+        )
+        rjson = resp.json()
+        result = rjson.get('result', {})
+        city = result['admin_district'] if 'admin_district' in result else None
+        return (city, None)
+    elif (
+        standardized_country_code == 'US'
+        or standardized_country_code == 'CA'
+        or standardized_country_code == 'fr'
+    ):
+        # FYI: Other countries Ziptastic supports:
+        # https://www.getziptastic.com/faq
+        try:
+            resp = requests.get(
+                f'{BASE_ZIPTASTIC_API_URL}{standardized_country_code}/{postal_code}',
+                headers = get_api_headers('ziptastic')
+            )
+            rjson = resp.json()
+        except:
+            return (None, None)
+
+        city = rjson.get('city', '')
+        state = rjson['state_short'] if 'state_short' in rjson else None
```

A more general potential issue, I'm noticing the pattern of returning None or empty string in the case that we aren't finding the data we need from the response. I think that's okay under the assumption we would only get back 200, or 404 responses. In that case, we've assumed that the API tells us it doesn't exist.

However, if we get something else, like a 5xx error or say the API changed and our request isn't fulfilled, then we would be returning None for what could be an otherwise valid postal code. Meaning we might construct Address objects with erroneously missing cities/states which would impact reporting or anywhere else we're using the API provided cities/states.

If it's appropriate we might want to throw an exception for non 200/404 statuses.

Additionally, we're handling a request exception for Ziptastic's API but are not doing the same for the postcodes API. Depending on what we go with from the above, letting the exception get  thrown might be the ideal.